### PR TITLE
Implementation of the OpenCL types

### PR DESCRIPTION
**Pull request by a-doumoulakis**
This pull request fixes in full or in part issues #41 #40 #16 

Here are the changes : 

- Implementation of the ```&&``` and ```||``` operations for ```small_array``` and all types inheriting from ```small_array``` 
- Implementation with Boost Preprocessor of most cl_types (```cl_char, cl_uchar, cl_short, cl_ushort, cl_int, cl_uint, cl_long, cl_ulong, cl_float, cl_double```) with size of 1,2,3,4,8 and 16. They can be used with or without OpenCL interoperability mode
- New test file that with the help of Boost Preprocessor tests all operators defined in small array (```* / % + - << >> < <= >= > == != &  | ^ && || ```) with all cl_types of all sizes

### CODE REVIEW

File: include/CL/sycl/opencl_type.hpp [hunk 1]
```diff
@@ -1,80 +1,69 @@
 #ifndef TRISYCL_SYCL_OPENCL_TYPE_HPP
 #define TRISYCL_SYCL_OPENCL_TYPE_HPP
 
-/** \file
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: Even if almost everything changed, you could keep the previous header

File: include/CL/sycl/opencl_type.hpp [hunk 2]
```diff
@@ -1,80 +1,69 @@
...
-/*
-    # Some Emacs stuff:
-    ### Local Variables:
-    ### ispell-local-dictionary: "american"
-    ### eval: (flyspell-prog-mode)
-    ### End:
-*/
-
-#endif //  TRISYCL_SYCL_OPENCL_TYPE_HPP
+#define SCALAR_TYPES                             \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: You should put all these preprocessor names in a more private "namespace" to limit collision probability with user code.
For example `TRISYCL__SCALAR_TYPES`

File: include/CL/sycl/opencl_type.hpp [hunk 3]
```diff
@@ -1,80 +1,69 @@
...
-    Joan DOT Thibault AT ens-rennes DOT fr
-
-    This file is distributed under the University of Illinois Open Source
-    License. See LICENSE.TXT for details.
-*/
-
-// Only provide OpenCL interoperability types when OpenCL is in use
-#ifdef TRISYCL_OPENCL
-
-#include <boost/compute.hpp>
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: At some point a `cl::sycl::cl_float3` for example has to be connected to an OpenCL `::cl_float3`, with some subtleties such as a `cl::sycl::cl_float3` as actually 4 elements and not 3...

There was recently some similar issue in OpenCL committee https://gitlab.khronos.org/opencl/headers/issues/6

File: include/CL/sycl/opencl_type.hpp [hunk 4]
```diff
@@ -1,80 +1,69 @@
...
-  {}
-
-
-  cl_float3 (float x, float y, float z) : self { x, y, z }
-  {}
-
-
-  /* Return the first element of the vector
-   */
-  auto& x() {
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: I guess removing this should break the ParallelSTL validation.

This reminds me that travis-CI should do more validation...

**Comment by keryell** [comment]
This reminds me another PR https://bitbucket.org/eigen/eigen/pull-requests/296/swizzle-operators-for-eigen

More generally, all the swizzle stuff has to be implemented some day...
For some references:
https://www.khronos.org/registry/OpenCL/specs/opencl-2.2-cplusplus.html#swizzle-methods
https://cvs.khronos.org/bugzilla/show_bug.cgi?id=14490

**Comment by keryell** [comment]
Does not compile for me.

**Comment by a-doumoulakis** [comment]
Sorry about that, I did not mean to push it. I still have to make some changes to make it work.

File: include/CL/sycl/opencl_type.hpp [hunk 5]
```diff
@@ -5,70 +5,196 @@
...
-#include <boost/compute.hpp>
 
 namespace cl {
 namespace sycl {
 
-/** Implement OpenCL interoperability types
-
-    \todo Implement all of them
+// List of the cl_types that will be iterated upon to generate the vector types
+#define SCALAR_TYPES                                                           \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_SCALAR_TYPES ?

File: include/CL/sycl/opencl_type.hpp [hunk 6]
```diff
@@ -5,70 +5,196 @@
...
+      ( uint   ,cl_uint,   unsigned int),                                      \
+      ( long   ,cl_long,   long int),                                          \
+      ( ulong  ,cl_ulong,  unsigned long int),                                 \
+      ( float  ,cl_float,  float),                                             \
+      ( double ,cl_double, double)                                             \
+      )                                                                        \
+    )
+
+// Accessors to get the type name, the cl_type name and the actual type
+#define TYPE_NAME(T)         BOOST_PP_TUPLE_ELEM(3, 0, T)
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_TYPE_NAME, TRISYCL_TYPE_CL_NAME, TRISYCL_TYPE_ACTUAL_NAME...

File: include/CL/sycl/opencl_type.hpp [hunk 7]
```diff
@@ -5,70 +5,196 @@
...
+#define TYPE_CL_NAME(T)      BOOST_PP_TUPLE_ELEM(3, 1, T)
+#define TYPE_ACTUAL_NAME(T)  BOOST_PP_TUPLE_ELEM(3, 2, T)
+
+// Return the name of the type concatenated with the size unless the size is 1
+#define SIZED_NAME(T, size)                      \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_SIZED_NAME...

File: include/CL/sycl/opencl_type.hpp [hunk 8]
```diff
@@ -5,70 +5,196 @@
...
+  BOOST_PP_IF(                                   \
+    BOOST_PP_EQUAL(size, 1), T,                  \
+    BOOST_PP_CAT(T, size))
+
+#ifndef TRISYCL_OPENCL
+
+/* If we are not using Boost Compute we just every cl_type with its \c vec
+   equivalent
 */
+#define DEFINE_TYPES(scalar, i)                  \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_DEFINE_TYPES

File: include/CL/sycl/opencl_type.hpp [hunk 9]
```diff
@@ -5,70 +5,196 @@
...
-
-
-  /* Return the third element of the vector
-   */
-  auto& z() {
-    return self.s[2];
-  }
-
-};
+#define IS_WRAPPER_TRAIT(type)                                                 \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_IS_WRAPPER_TRAIT

File: include/CL/sycl/opencl_type.hpp [hunk 10]
```diff
@@ -5,70 +5,196 @@
...
+    BOOST_PP_EQUAL(size, 1), T,                  \
+    BOOST_PP_CAT(T, size))
+
+#ifndef TRISYCL_OPENCL
+
+/* If we are not using Boost Compute we just every cl_type with its \c vec
+   equivalent
 */
+#define DEFINE_TYPES(scalar, i)                  \
+  using SIZED_NAME(TYPE_CL_NAME(scalar), i) =    \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 11]
```diff
@@ -5,70 +5,196 @@
...
+template <> struct is_wrapper<type> : std::true_type {};
 
-}
-}
+/* We define 3 different wrapper classes around OpenCL types.
+   These classes allow us to use t.x() and t.y() for vector types of size 2.
+   And in addition we can also use t.z() and t.w() for vector types of size
+   3 and 4 similarly to OpenCL vector types
+*/
+#define WRAPPER_CLASS_2(cl_type, boost_name, scalar_type)                      \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 12]
```diff
@@ -5,70 +5,196 @@
...
+  class cl_type {                                                              \
+    ::cl_type self;                                                            \
+                                                                               \
+  public:                                                                      \
+    cl_type () = default;                                                      \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: Is the space required by the preprocessor between cl_type and ()?

File: include/CL/sycl/opencl_type.hpp [hunk 13]
```diff
@@ -5,70 +5,196 @@
...
+    cl_type (::cl_type self_) : self { self_ } {}                              \
+    cl_type (scalar_type x, scalar_type y) : self { x, y } {}                  \
+    auto& x() { return self.s[0]; }                                            \
+    auto& y() { return self.s[1]; }                                            \
+    auto& unwrap() const { return self; }};                                    \
+  IS_WRAPPER_TRAIT(cl_type)
+
+#define WRAPPER_CLASS_3(cl_type, boost_name, scalar_type)                      \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 14]
```diff
@@ -5,70 +5,196 @@
...
+    cl_type (::cl_type self_) : self { self_ } {}                              \
+    cl_type (scalar_type x, scalar_type y, scalar_type z) :                    \
+      self { x, y, z } {}                                                      \
+    auto& x() { return self.s[0]; }                                            \
+    auto& y() { return self.s[1]; }                                            \
+    auto& z() { return self.s[2]; }                                            \
+    auto& unwrap() const { return self; }};                                    \
+  IS_WRAPPER_TRAIT(cl_type)
+
+#define WRAPPER_CLASS_4(cl_type, boost_name, scalar_type)                      \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 15]
```diff
@@ -5,70 +5,196 @@
...
+    auto& x() { return self.s[0]; }                                            \
+    auto& y() { return self.s[1]; }                                            \
+    auto& z() { return self.s[2]; }                                            \
+    auto& w() { return self.s[3]; }                                            \
+    auto& unwrap() const { return self; }};                                    \
+  IS_WRAPPER_TRAIT(cl_type)
+
+
+// Return the Boost Compute type for OpenCL vector types
+#define BOOST_COMPUTE_NAME(scalar, size)                                       \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 16]
```diff
@@ -5,70 +5,196 @@
...
+  SIZED_NAME(boost::compute::TYPE_NAME(scalar), size)
+
+/* For vector types of size above 4 and 1 we typedef the vector type with
+   its Boost Compute equivalent
+*/
+#define TYPEDEF_TYPE(cl_type, boost_name, scalar_type)                         \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 17]
```diff
@@ -5,70 +5,196 @@
...
+  using cl_type = boost_name##_;
+
+// Helper macro to properly define each OpenCL type
+#define P_DEFINE_TYPE(cl_type, boost_name, scalar_type, i)                     \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 18]
```diff
@@ -5,70 +5,196 @@
...
+              BOOST_PP_IF(BOOST_PP_EQUAL(i, 2), WRAPPER_CLASS_2,               \
+                          BOOST_PP_IF(BOOST_PP_EQUAL(i, 3),                    \
+                                      WRAPPER_CLASS_3,                         \
+                                      BOOST_PP_IF(BOOST_PP_EQUAL(i, 4),        \
+                                                  WRAPPER_CLASS_4,             \
+                                                  TYPEDEF_TYPE))))             \
+  (cl_type, boost_name, scalar_type)
+
+// Helper macro to properly define each OpenCL type
+#define DEFINE_TYPES(scalar, i)                                                \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 19]
```diff
@@ -5,70 +5,196 @@
...
+  P_DEFINE_TYPE(SIZED_NAME(TYPE_CL_NAME(scalar), i),                           \
+                BOOST_COMPUTE_NAME(scalar, i),                                 \
+                TYPE_ACTUAL_NAME(scalar), i)
 
 #endif
 
+/* We declare vector types of sizes 1,2,3,4,8 and 16 as per
+   the SYCL specification
+*/
+#define DECLARE_CL_TYPES(r, data, scalar)            \
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: TRISYCL_

File: include/CL/sycl/opencl_type.hpp [hunk 20]
```diff
@@ -5,70 +5,196 @@
...
+
+// Generate the vector types for all listed scalar types
+BOOST_PP_LIST_FOR_EACH(DECLARE_CL_TYPES, _, SCALAR_TYPES)
+
+} // sycl
+} // cl
+
+
+// Undef macros to avoid name collision
+#undef SCALAR_TYPES
```
Comment (keryell) [inline include/CL/sycl/opencl_type.hpp]: To update accordingly

File: include/CL/sycl/vec.hpp [hunk 21]
```diff
@@ -152,8 +152,9 @@ class vec : public detail::small_array<DataType,
 #endif
 };
 
-  /** A macro to define type alias, such as for type=uchar, size=4 and
-      real_type=unsigned char, uchar4 is equivalent to vec<float, 4>
+
+ /** A macro to define type alias, such as for type=uchar, size=4 and
```
Comment (keryell) [inline include/CL/sycl/vec.hpp]: Lacking space in the indentation

File: tests/kernel/opencl_kernel_vector_add_args_42.cpp [hunk 22]
```diff
@@ -54,9 +54,7 @@ int test_main(int argc, char *argv[]) {
         cgh.set_args(A.get_access<access::mode::read>(cgh),
                      B.get_access<access::mode::read>(cgh),
                      C.get_access<access::mode::write>(cgh),
-                     /* TODO: use cl::sycl::cl_int { 42 } for portability
-                        when it is implemented */
-                     42);
+                     cl::sycl::cl_int { 42 } );
```
Comment (keryell) [inline tests/kernel/opencl_kernel_vector_add_args_42.cpp]: I love that ! :-)

**Comment by keryell** [comment]
Thank you for your contribution!

**Pull request merged**